pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/SegmentCommitterFactory.java

```diff
@@ -69,9 +70,13 @@ public SegmentCommitter createSegmentCommitter(boolean isSplitCommit, SegmentCom
 
     boolean uploadToFs = _streamConfig.isServerUploadToDeepStore();
     String peerSegmentDownloadScheme = _tableConfig.getValidationConfig().getPeerSegmentDownloadScheme();
-    // TODO: exists for backwards compatibility. remove peerDownloadScheme non-null check once users have migrated
-    if (uploadToFs || peerSegmentDownloadScheme != null) {
-      segmentUploader = new PinotFSSegmentUploader(_indexLoadingConfig.getSegmentStoreURI(),
+    String segmentStoreUri = _indexLoadingConfig.getSegmentStoreURI();
+
+    // We seem to allow the server instance to come up without a valid segment store uri. Hence, this check is needed.
+    if (!Strings.isNullOrEmpty(segmentStoreUri)
```

Comment:
(minor) We usually use `StringUtils.isEmpty()` (from `org.apache.commons.lang3`)

Reply:
(MAJOR) When segment upload fails, but peer download is enabled, we still want to use the `PinotFSSegmentUploader` because peer download doesn't require segment to be successfully uploaded. We can attach this check to the `uploadToFs`

Reply:
> (MAJOR) When segment upload fails, but peer download is enabled, we still want to use the `PinotFSSegmentUploader` because peer download doesn't require segment to be successfully uploaded. We can attach this check to the `uploadToFs`

This PR doesn't really change anything with respect to upload failure handling. These configs only control what uploader is being used. If you see SplitSegmentCommitter [here](https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/SplitSegmentCommitter.java#L99), this part handles the upload failure in the presence of a peer download scheme. The behavior is the same as it was before this PR. 

Note: It might be cleaner to have a failure handler provided as the committer context. but that still doesn't change any existing behavior

Reply:
Fixed based on our offline discussion. Modified to not default to controller upload. Instead added config validation before table data manager is created.